### PR TITLE
Feature settings 

### LINKED ISSUE #472: Settings page
Need a setting page. Probably with just a json file editor for the moment

### PR DESCRIPTION
**Pull request by timotheeguerin**
fix #472 
image

### CODE REVIEW

**Comment by paselem** [comment]
This is missing a milestone.

File: app/components/base/editor/editor.component.ts [hunk 1]
```diff
@@ -83,6 +95,10 @@ export class EditorComponent implements ControlValueAccessor, AfterViewInit, OnC
             this.blur.emit();
             this.changeDetector.markForCheck();
         });
+
+        setTimeout(() => {
```
Comment (paselem) [inline app/components/base/editor/editor.component.ts]: Shouldn't you be discarding this timeout when you navigate away from the component?

File: app/components/file/details/file-content.component.ts [hunk 2]
```diff
@@ -1,19 +1,20 @@
...
-const fileTypes = {
-    [FileType.log]: ["txt", "log"],
-    [FileType.image]: ["png", "jpg", "gif"],
-    [FileType.code]: ["json", "ts2", "js", "java", "cs", "cpp", "h", "hpp", "py", "xml", "sh", "cmd", "bat"],
-};
+// const fileTypes = {
+//     [FileType.log]: ["txt", "log"],
+//     [FileType.image]: ["png", "jpg", "gif"],
+//     [FileType.code]: ["json", "ts2", "js", "java", "cs", "cpp", "h", "hpp", "py", "xml", "sh", "cmd", "bat"],
+// };
```
Comment (paselem) [inline app/components/file/details/file-content.component.ts]: Remove?

File: app/components/settings/settings.component.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        readOnly: true,
+        lineNumbers: true,
+        mode: "application/json",
+        tabSize: 2,
+        indentUnit: 2,
+    };
+
+    public userSettingsEditorConfig: CodeMirror.EditorConfiguration = {
+        lineNumbers: true,
+        mode: "application/javascript",
```
Comment (paselem) [inline app/components/settings/settings.component.ts]: Why are we using json above, but javascript here?

File: app/components/settings/settings.component.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        tabSize: 2,
+        indentUnit: 2,
+        gutters: ["CodeMirror-lint-markers"],
+        lint: true,
+        extraKeys: {
+            "Ctrl-S": this.save,
```
Comment (paselem) [inline app/components/settings/settings.component.ts]: I saw above you were adding Ctrl-/ in code. Should it be added here?

File: app/components/settings/settings.component.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        this._originalUserSettings = str;
+
+        if (this.userSettings.untouched) {
+            this.userSettings.setValue(str);
+        }
+    }
+
+    @autobind()
+    private _validJsonConfig(c: FormControl): Promise<any> {
+        clearTimeout(this._validateDebounceTimeout);
```
Comment (paselem) [inline app/components/settings/settings.component.ts]: does clearTimeout() care if this._validateDebounceTimeout is undefined?

File: app/models/settings.ts [hunk 6]
```diff
@@ -1,8 +1,8 @@
-export interface Settings {
-    homePage: "account" | "home" | "job";
+export enum Theme {
```
Comment (paselem) [inline app/models/settings.ts]: While this looks nice to have in the settings, is it worth exposing if we only have 1??

File: app/utils/extensions/json.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+interface RegExpConstructor {
+    leftContext: string;
+    rightContext: string;
+}
+
+if (!JSON.minify) {
+    JSON.minify = (json) => {
+
+        let tokenizer = /"|(\/\*)|(\*\/)|(\/\/)|\n|\r/g;
```
Comment (paselem) [inline app/utils/extensions/json.ts]: ??? What? Please add some comments describing what this does.

File: app/components/settings/settings.component.ts (hunk 5 above)
Comment (timotheeguerin) [inline app/components/settings/settings.component.ts]: nope

File: app/utils/extensions/json.ts (hunk 7 above)
Comment (timotheeguerin) [inline app/utils/extensions/json.ts]: removed it and used `strip-json-comments` library

File: app/models/settings.ts (hunk 6 above)
Comment (timotheeguerin) [inline app/models/settings.ts]: I guess we can remove it

File: app/components/settings/settings.component.ts (hunk 4 above)
Comment (timotheeguerin) [inline app/components/settings/settings.component.ts]: yeah ctrl+/ is added for all language but the save depends on the usage of the editor

**Pull request merged**